src/typechecker.rs

```diff
@@ -34,6 +34,83 @@ impl Display for CheckedType {
     }
 }
 
+// TODO: Should we factor this into a `Context` trait? All contexts will share some
+// similarities, such as the ability to emit errors or enter and exit scopes
+/// The [`TypeCtx`]'s role is to keep track of declared types based on the scope they
+/// were created in. In doing so, it allows the [`Instruction`]s to perform a type lookup
+/// in order to resolve to a concrete type.Each declaration (First [`VarAssign`],
+/// [`FunctionDec`]s and [`TypeDec`]s) can also declare a new type and make it available
+/// to all instructions in the avaialble scopes.
+pub struct TypeCtx<'ctx> {
```

Comment:
What's the difference between a Scope and a Context or a TypeContext ?

Reply:
A Scope keeps track of all variables, functions and types declared in the current lexical scope. A Context is responsible for executing code, typechecking it and later optimizing it (and maybe compiling it etc). Basically, it's the entry point of the interpreter.
A TypeContext is a "subcontext" responsible only for declaring types and resolving them later on. Basically, a FuncDec will declare its type, and function calls can later on fetch that type to make sure everything fits properly